File: hallucinate_app/node/github_issue_reporter.js
/**
 * GitHub Issue Reporter (JavaScript)
 * 
 * Automatically creates GitHub issues from runtime errors.
 * Integrates with error handling in Node.js/Electron applications.
 */

import crypto from 'crypto';
import fs from 'fs';
import path from 'path';

/**
 * Configuration for GitHub issue reporting
 */
class IssueReportConfig {
  constructor(options = {}) {
    this.enabled = options.enabled ?? (process.env.GITHUB_ISSUE_REPORTER_ENABLED === 'true');
    this.githubToken = options.githubToken ?? process.env.GITHUB_TOKEN;
    this.repository = options.repository ?? process.env.GITHUB_REPOSITORY;
    this.minErrorLevel = options.minErrorLevel ?? process.env.GITHUB_ISSUE_MIN_LEVEL ?? 'error';
    this.rateLimitPerHour = options.rateLimitPerHour ?? parseInt(process.env.GITHUB_ISSUE_RATE_LIMIT || '10');
    this.duplicateCheckWindowHours = options.duplicateCheckWindowHours ?? parseInt(process.env.GITHUB_ISSUE_DUPLICATE_WINDOW || '24');
    this.issueLabels = options.issueLabels ?? (process.env.GITHUB_ISSUE_LABELS || 'auto-reported,bug').split(',');
    this.assignees = options.assignees ?? (process.env.GITHUB_ISSUE_ASSIGNEES || '').split(',').filter(a => a);
    this.autoAssign = options.autoAssign ?? (process.env.GITHUB_ISSUE_AUTO_ASSIGN === 'true');
    this.includeStackTrace = options.includeStackTrace ?? (process.env.GITHUB_ISSUE_INCLUDE_STACK !== 'false');
    this.includeSystemInfo = options.includeSystemInfo ?? (process.env.GITHUB_ISSUE_INCLUDE_SYSTEM_INFO !== 'false');
    this.dryRun = options.dryRun ?? (process.env.GITHUB_ISSUE_DRY_RUN === 'true');
  }
}

/**
 * Error levels enum
 */
const ErrorLevel = {
  DEBUG: 'debug',
  INFO: 'info',
  WARNING: 'warning',
  ERROR: 'error',
  FATAL: 'fatal'
};

/**
 * Error sources enum
 */
const ErrorSource = {
  JAVASCRIPT: 'javascript',
  NODEJS: 'nodejs',
  ELECTRON: 'electron',
  MCP_SERVER: 'mcp-server',
  DASHBOARD: 'dashboard',
  UNKNOWN: 'unknown'
};

/**
 * GitHub Issue Reporter
 * 
 * Features:
 * - Duplicate detection using error fingerprinting
 * - Rate limiting to prevent API abuse
 * - Configurable severity filtering
 * - Rich issue formatting with stack traces
 */
class GitHubIssueReporter {
  constructor(config = null) {
    this.config = config || new IssueReportConfig();
    
    // Track reported issues to prevent duplicates
    this.reportedFingerprints = new Set();
    this.fingerprintTimestamps = new Map();
    
    // Rate limiting
    this.issueCountPerHour = [];
    
    // Statistics
    this.stats = {
      totalErrorsProcessed: 0,
      issuesCreated: 0,
      duplicatesSkipped: 0,
      rateLimited: 0,
      errors: 0
    };
    
    // Initialize Octokit if available
    this.octokit = null;
    if (this.config.enabled) {
      this._initOctokit();
    }
  }
  
  /**
   * Initialize the Octokit GitHub API client
   * @private
   */
  async _initOctokit() {
    if (!this.config.githubToken) {
      console.error('GitHub token not configured');
      return;
    }
    
    if (!this.config.repository) {
      console.error('GitHub repository not configured');
      return;
    }
    
    try {
      // Dynamic import of @octokit/rest
      const { Octokit } = await import('@octokit/rest');
      
      this.octokit = new Octokit({
        auth: this.config.githubToken
      });
      
      // Parse repository
      const [owner, repo] = this.config.repository.split('/');
      this.owner = owner;
      this.repo = repo;
      
      console.log(`GitHub client initialized for repository: ${this.config.repository}`);
    } catch (error) {
      console.error('Failed to initialize GitHub client:', error.message);
      console.error('Install @octokit/rest with: npm install @octokit/rest');
      this.octokit = null;
    }
  }
  
  /**
   * Generate a unique fingerprint for an error
   * @param {Object} error - Error data
   * @returns {string} SHA256 hash of error characteristics
   */
  generateErrorFingerprint(error) {
    // Create fingerprint from error type, component, and message
    const fingerprintData = `${error.level}:${error.source}:${error.component}:${error.message.substring(0, 200)}`;
    return crypto.createHash('sha256').update(fingerprintData).digest('hex').substring(0, 16);
  }
  
  /**
   * Check if an error with this fingerprint was recently reported
   * @param {string} fingerprint - Error fingerprint
   * @returns {boolean} True if duplicate
   */
  isDuplicate(fingerprint) {
    if (!this.fingerprintTimestamps.has(fingerprint)) {
      return false;
    }
    
    // Check if the fingerprint is within the duplicate check window
    const lastReported = this.fingerprintTimestamps.get(fingerprint);
    const windowMs = this.config.duplicateCheckWindowHours * 3600 * 1000;
    
    if (Date.now() - lastReported < windowMs) {
      return true;
    }
    
    // Clean up old fingerprint
    this.fingerprintTimestamps.delete(fingerprint);
    this.reportedFingerprints.delete(fingerprint);
    return false;
  }
  
  /**
   * Check if we've hit the rate limit
   * @returns {boolean} True if within rate limit, False if exceeded
   */
  checkRateLimit() {
    // Clean up old timestamps (older than 1 hour)
    const now = Date.now();
    this.issueCountPerHour = this.issueCountPerHour.filter(
      ts => now - ts < 3600000
    );
    
    return this.issueCountPerHour.length < this.config.rateLimitPerHour;
  }
  
  /**
   * Format the issue title
   * @param {Object} error - Error data
   * @returns {string} Formatted issue title
   */
  formatIssueTitle(error) {
    // Extract error type from metadata or name
    const errorType = error.metadata?.errorType || error.name || 'Error';
    
    // Create concise title
    let title = `[${error.level.toUpperCase()}] ${errorType} in ${error.component}`;
    
    // Truncate if too long
    if (title.length > 80) {
      title = title.substring(0, 77) + '...';
    }
    
    return title;
  }
  
  /**
   * Format the issue body with error details
   * @param {Object} error - Error data
   * @returns {string} Formatted issue body in Markdown
   */
  formatIssueBody(error) {
    const lines = [];
    
    // Header
    lines.push('## Auto-Generated Error Report');
    lines.push('');
    lines.push(`**Error ID:** \`${error.id}\``);
    lines.push(`**Timestamp:** ${error.timestamp}`);
    lines.push(`**Level:** \`${error.level}\``);
    lines.push(`**Source:** \`${error.source}\``);
    lines.push(`**Component:** \`${error.component}\``);
    lines.push(`**Operation:** \`${error.operation}\``);
    lines.push('');
    
    // Error message
    lines.push('### Error Message');
    lines.push('```');
    lines.push(error.message);
    lines.push('```');
    lines.push('');
    
    // Stack trace
    if (this.config.includeStackTrace && error.stackTrace) {
      lines.push('### Stack Trace');
      lines.push('```');
      lines.push(error.stackTrace);
      lines.push('```');
      lines.push('');
    }
    
    // Details
    if (error.details && Object.keys(error.details).length > 0) {
      lines.push('### Additional Details');
      lines.push('```json');
      lines.push(JSON.stringify(error.details, null, 2));
      lines.push('```');
      lines.push('');
    }
    
    // System info
    if (this.config.includeSystemInfo && error.metadata && Object.keys(error.metadata).length > 0) {
      lines.push('### System Information');
      for (const [key, value] of Object.entries(error.metadata)) {
        lines.push(`- **${key}:** \`${value}\``);
      }
      lines.push('');
    }
    
    // Tags
    if (error.tags && error.tags.length > 0) {
      lines.push('### Tags');
      lines.push(error.tags.map(tag => `\`${tag}\``).join(', '));
      lines.push('');
    }
    
    // Occurrence info
    if (error.count && error.count > 1) {
      lines.push('### Occurrence Information');
      lines.push(`- **Total occurrences:** ${error.count}`);
      if (error.firstSeen) lines.push(`- **First seen:** ${error.firstSeen}`);
      if (error.lastSeen) lines.push(`- **Last seen:** ${error.lastSeen}`);
      lines.push('');
    }
    
    // Footer
    lines.push('---');
    lines.push('*This issue was automatically generated by the error monitoring system.*');
    
    return lines.join('\n');
  }
  
  /**
   * Determine if an error should be reported to GitHub
   * @param {Object} error - Error data
   * @returns {boolean} True if error should be reported
   */
  shouldReportError(error) {
    // Check if reporting is enabled
    if (!this.config.enabled) {
      return false;
    }
    
    // Check if GitHub client is initialized
    if (!this.octokit) {
      return false;
    }
    
    // Check error level
    const errorLevels = [ErrorLevel.DEBUG, ErrorLevel.INFO, ErrorLevel.WARNING, ErrorLevel.ERROR, ErrorLevel.FATAL];
    const minLevelIndex = errorLevels.indexOf(this.config.minErrorLevel);
    const errorLevelIndex = errorLevels.indexOf(error.level);
    
    if (errorLevelIndex < minLevelIndex) {
      return false;
    }
    
    // Check for duplicates
    const fingerprint = this.generateErrorFingerprint(error);
    if (this.isDuplicate(fingerprint)) {
      this.stats.duplicatesSkipped++;
      console.debug(`Skipping duplicate error: ${fingerprint}`);
      return false;
    }
    
    // Check rate limit
    if (!this.checkRateLimit()) {
      this.stats.rateLimited++;
      console.warn('GitHub issue rate limit exceeded');
      return false;
    }
    
    return true;
  }
  
  /**
   * Create a GitHub issue for the error
   * @param {Object} error - Error data
   * @returns {Promise<string|null>} Issue URL if created, null otherwise
   */
  async createIssue(error) {
    this.stats.totalErrorsProcessed++;
    
    // Check if we should report this error
    if (!this.shouldReportError(error)) {
      return null;
    }
    
    try {
      // Format issue
      const title = this.formatIssueTitle(error);
      const body = this.formatIssueBody(error);
      
      // Dry run mode
      if (this.config.dryRun) {
        console.log(`DRY RUN - Would create issue: ${title}`);
        console.debug(`Issue body:\n${body}`);
        this.stats.issuesCreated++;
        return 'dry-run-issue-url';
      }
      
      // Create issue
      const response = await this.octokit.rest.issues.create({
        owner: this.owner,
        repo: this.repo,
        title: title,
        body: body,
        labels: this.config.issueLabels,
        assignees: this.config.autoAssign ? this.config.assignees : undefined
      });
      
      // Track this issue
      const fingerprint = this.generateErrorFingerprint(error);
      this.reportedFingerprints.add(fingerprint);
      this.fingerprintTimestamps.set(fingerprint, Date.now());
      this.issueCountPerHour.push(Date.now());
      
      this.stats.issuesCreated++;
      
      console.log(`Created GitHub issue: ${response.data.html_url}`);
      return response.data.html_url;
      
    } catch (error) {
      this.stats.errors++;
      console.error(`Failed to create GitHub issue:`, error);
      return null;
    }
  }
  
  /**
   * Get reporter statistics
   * @returns {Object} Statistics
   */
  getStats() {
    return {
      ...this.stats,
      config: {
        enabled: this.config.enabled,
        repository: this.config.repository,
        minErrorLevel: this.config.minErrorLevel,
        rateLimitPerHour: this.config.rateLimitPerHour,
        dryRun: this.config.dryRun
      },
      currentRate: this.issueCountPerHour.length,
      trackedFingerprints: this.reportedFingerprints.size
    };
  }
  
  /**
   * Create an error object from an Error instance
   * @param {Error} err - JavaScript Error object
   * @param {Object} options - Additional options
   * @returns {Object} Error data object
   */
  static createErrorFromException(err, options = {}) {
    const {
      component = 'unknown',
      operation = 'unknown',
      source = ErrorSource.JAVASCRIPT,
      level = ErrorLevel.ERROR,
      details = {}
    } = options;
    
    return {
      id: crypto.randomUUID(),
      timestamp: new Date().toISOString(),
      level: level,
      source: source,
      component: component,
      operation: operation,
      message: err.message,
      name: err.name,
      stackTrace: err.stack,
      details: details,
      metadata: {
        errorType: err.name,
        nodeVersion: process.version,
        platform: process.platform,
        arch: process.arch
      },
      tags: [component.toLowerCase(), source, level],
      count: 1
    };
  }
}

// Singleton instance
let reporterInstance = null;

/**
 * Get the singleton GitHub issue reporter instance
 * @param {IssueReportConfig} config - Optional configuration (only used on first call)
 * @returns {GitHubIssueReporter} The reporter instance
 */
export function getReporter(config = null) {
  if (!reporterInstance) {
    reporterInstance = new GitHubIssueReporter(config);
  }
  return reporterInstance;
}

/**
 * Convenience function to report an error
 * @param {Object} error - Error data
 * @returns {Promise<string|null>} Issue URL if created
 */
export async function reportError(error) {
  const reporter = getReporter();
  return await reporter.createIssue(error);
}

/**
 * Convenience function to report an Error exception
 * @param {Error} err - JavaScript Error object
 * @param {Object} options - Additional options
 * @returns {Promise<string|null>} Issue URL if created
 */
export async function reportException(err, options = {}) {
  const error = GitHubIssueReporter.createErrorFromException(err, options);
  return await reportError(error);
}

export { GitHubIssueReporter, IssueReportConfig, ErrorLevel, ErrorSource };
export default GitHubIssueReporter;

File: hallucinate_app/python/hallucinate_app/github_issue_reporter.py
"""
GitHub Issue Reporter

Automatically creates GitHub issues from runtime errors.
Integrates with the ErrorMonitor system to report errors to GitHub Issues.
"""

import os
import json
import hashlib
import logging
import time
from typing import Dict, List, Optional, Any, Set
from datetime import datetime, timedelta
from dataclasses import dataclass, field

# Set up logging
logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s - %(name)s - %(levelname)s - %(message)s"
)
logger = logging.getLogger("github_issue_reporter")

try:
    from github import Github, GithubException
    GITHUB_AVAILABLE = True
except ImportError:
    GITHUB_AVAILABLE = False
    logger.warning("PyGithub not available. Install with: pip install PyGithub")

from hallucinate_app.error_monitor import ErrorData, ErrorLevel, ErrorSource


@dataclass
class IssueReportConfig:
    """Configuration for GitHub issue reporting"""
    enabled: bool = False
    github_token: Optional[str] = None
    repository: Optional[str] = None  # Format: "owner/repo"
    min_error_level: ErrorLevel = ErrorLevel.ERROR
    rate_limit_per_hour: int = 10
    duplicate_check_window_hours: int = 24
    issue_labels: List[str] = field(default_factory=lambda: ["auto-reported", "bug"])
    assignees: List[str] = field(default_factory=list)
    auto_assign: bool = False
    include_stack_trace: bool = True
    include_system_info: bool = True
    dry_run: bool = False  # If True, log instead of creating issues
    
    @classmethod
    def from_env(cls) -> 'IssueReportConfig':
        """Create configuration from environment variables"""
        return cls(
            enabled=os.getenv('GITHUB_ISSUE_REPORTER_ENABLED', 'false').lower() == 'true',
            github_token=os.getenv('GITHUB_TOKEN'),
            repository=os.getenv('GITHUB_REPOSITORY'),
            min_error_level=ErrorLevel(os.getenv('GITHUB_ISSUE_MIN_LEVEL', 'error')),
            rate_limit_per_hour=int(os.getenv('GITHUB_ISSUE_RATE_LIMIT', '10')),
            duplicate_check_window_hours=int(os.getenv('GITHUB_ISSUE_DUPLICATE_WINDOW', '24')),
            issue_labels=os.getenv('GITHUB_ISSUE_LABELS', 'auto-reported,bug').split(','),
            assignees=os.getenv('GITHUB_ISSUE_ASSIGNEES', '').split(',') if os.getenv('GITHUB_ISSUE_ASSIGNEES') else [],
            auto_assign=os.getenv('GITHUB_ISSUE_AUTO_ASSIGN', 'false').lower() == 'true',
            include_stack_trace=os.getenv('GITHUB_ISSUE_INCLUDE_STACK', 'true').lower() == 'true',
            include_system_info=os.getenv('GITHUB_ISSUE_INCLUDE_SYSTEM_INFO', 'true').lower() == 'true',
            dry_run=os.getenv('GITHUB_ISSUE_DRY_RUN', 'false').lower() == 'true',
        )


class GitHubIssueReporter:
    """
    Automatically creates GitHub issues from errors.
    
    Features:
    - Duplicate detection using error fingerprinting
    - Rate limiting to prevent API abuse
    - Configurable severity filtering
    - Rich issue formatting with stack traces
    """
    
    def __init__(self, config: Optional[IssueReportConfig] = None):
        """
        Initialize the GitHub issue reporter
        
        Args:
            config: Configuration for issue reporting
        """
        self.config = config or IssueReportConfig.from_env()
        
        # Initialize GitHub client
        self.github_client = None
        self.repo = None
        
        if self.config.enabled and GITHUB_AVAILABLE:
            self._init_github_client()
        
        # Track reported issues to prevent duplicates
        self.reported_fingerprints: Set[str] = set()
        self.fingerprint_timestamps: Dict[str, float] = {}
        
        # Rate limiting
        self.issue_count_per_hour: List[float] = []
        
        # Statistics
        self.stats = {
            "total_errors_processed": 0,
            "issues_created": 0,
            "duplicates_skipped": 0,
            "rate_limited": 0,
            "errors": 0
        }
        self.last_issue_error: Optional[Dict[str, Any]] = None
    
    def _init_github_client(self):
        """Initialize the GitHub API client"""
        if not self.config.github_token:
            logger.error("GitHub token not configured")
            return
        
        if not self.config.repository:
            logger.error("GitHub repository not configured")
            return
        
        try:
            self.github_client = Github(self.config.github_token)
            self.repo = self.github_client.get_repo(self.config.repository)
            logger.info(f"GitHub client initialized for repository: {self.config.repository}")
        except Exception as e:
            logger.exception(f"Failed to initialize GitHub client: {e}")
            self.github_client = None
            self.repo = None
    
    def generate_error_fingerprint(self, error: ErrorData) -> str:
        """
        Generate a unique fingerprint for an error
        
        Args:
            error: Error data
            
        Returns:
            str: SHA256 hash of error characteristics
        """
        # Create fingerprint from error type, component, and message
        fingerprint_data = f"{error.level.value}:{error.source.value}:{error.component}:{error.message[:200]}"
        return hashlib.sha256(fingerprint_data.encode()).hexdigest()[:16]
    
    def is_duplicate(self, fingerprint: str) -> bool:
        """
        Check if an error with this fingerprint was recently reported
        
        Args:
            fingerprint: Error fingerprint
            
        Returns:
            bool: True if duplicate
        """
        if fingerprint not in self.fingerprint_timestamps:
            return False
        
        # Check if the fingerprint is within the duplicate check window
        last_reported = self.fingerprint_timestamps[fingerprint]
        window_seconds = self.config.duplicate_check_window_hours * 3600
        
        if time.time() - last_reported < window_seconds:
            return True
        
        # Clean up old fingerprint
        del self.fingerprint_timestamps[fingerprint]
        self.reported_fingerprints.discard(fingerprint)
        return False
    
    def check_rate_limit(self) -> bool:
        """
        Check if we've hit the rate limit
        
        Returns:
            bool: True if within rate limit, False if exceeded
        """
        # Clean up old timestamps (older than 1 hour)
        now = time.time()
        self.issue_count_per_hour = [
            ts for ts in self.issue_count_per_hour 
            if now - ts < 3600
        ]
        
        return len(self.issue_count_per_hour) < self.config.rate_limit_per_hour
    
    def format_issue_title(self, error: ErrorData) -> str:
        """
        Format the issue title
        
        Args:
            error: Error data
            
        Returns:
            str: Formatted issue title
        """
        # Extract error type from message if available
        error_type = error.metadata.get('error_type', 'Error')
        
        # Create concise title
        title = f"[{error.level.value.upper()}] {error_type} in {error.component}"
        
        # Truncate if too long
        if len(title) > 80:
            title = title[:77] + "..."
        
        return title
    
    def format_issue_body(self, error: ErrorData) -> str:
        """
        Format the issue body with error details
        
        Args:
            error: Error data
            
        Returns:
            str: Formatted issue body in Markdown
        """
        lines = []
        
        # Header
        lines.append("## Auto-Generated Error Report")
        lines.append("")
        lines.append(f"**Error ID:** `{error.id}`")
        lines.append(f"**Timestamp:** {error.timestamp}")
        lines.append(f"**Level:** `{error.level.value}`")
        lines.append(f"**Source:** `{error.source.value}`")
        lines.append(f"**Component:** `{error.component}`")
        lines.append(f"**Operation:** `{error.operation}`")
        lines.append("")
        
        # Error message
        lines.append("### Error Message")
        lines.append("```")
        lines.append(error.message)
        lines.append("```")
        lines.append("")
        
        # Stack trace
        if self.config.include_stack_trace and error.stack_trace:
            lines.append("### Stack Trace")
            lines.append("```")
            lines.append(error.stack_trace)
            lines.append("```")
            lines.append("")
        
        # Details
        if error.details:
            lines.append("### Additional Details")
            lines.append("```json")
            lines.append(json.dumps(error.details, indent=2))
            lines.append("```")
            lines.append("")
        
        # System info
        if self.config.include_system_info and error.metadata:
            lines.append("### System Information")
            for key, value in error.metadata.items():
                lines.append(f"- **{key}:** `{value}`")
            lines.append("")
        
        # Tags
        if error.tags:
            lines.append("### Tags")
            lines.append(", ".join([f"`{tag}`" for tag in error.tags]))
            lines.append("")
        
        # Occurrence info
        if error.count > 1:
            lines.append("### Occurrence Information")
            lines.append(f"- **Total occurrences:** {error.count}")
            lines.append(f"- **First seen:** {error.first_seen}")
            lines.append(f"- **Last seen:** {error.last_seen}")
            lines.append("")
        
        # Footer
        lines.append("---")
        lines.append("*This issue was automatically generated by the error monitoring system.*")
        
        return "\n".join(lines)
    
    def should_report_error(self, error: ErrorData) -> bool:
        """
        Determine if an error should be reported to GitHub
        
        Args:
            error: Error data
            
        Returns:
            bool: True if error should be reported
        """
        # Check if reporting is enabled
        if not self.config.enabled:
            return False
        
        # Check if GitHub client is initialized
        if not self.repo:
            return False
        
        # Check error level
        error_levels = [ErrorLevel.DEBUG, ErrorLevel.INFO, ErrorLevel.WARNING, ErrorLevel.ERROR, ErrorLevel.FATAL]
        min_level_index = error_levels.index(self.config.min_error_level)
        error_level_index = error_levels.index(error.level)
        
        if error_level_index < min_level_index:
            return False
        
        # Check for duplicates
        fingerprint = self.generate_error_fingerprint(error)
        if self.is_duplicate(fingerprint):
            self.stats["duplicates_skipped"] += 1
            logger.debug(f"Skipping duplicate error: {fingerprint}")
            return False
        
        # Check rate limit
        if not self.check_rate_limit():
            self.stats["rate_limited"] += 1
            logger.warning("GitHub issue rate limit exceeded")
            return False
        
        return True
    
    def create_issue(self, error: ErrorData) -> Optional[str]:
        """
        Create a GitHub issue for the error
        
        Args:
            error: Error data
            
        Returns:
            Optional[str]: Issue URL if created, None otherwise
        """
        self.stats["total_errors_processed"] += 1
        
        # Check if we should report this error
        if not self.should_report_error(error):
            return None
        
        try:
            # Format issue
            title = self.format_issue_title(error)
            body = self.format_issue_body(error)
            
            # Dry run mode
            if self.config.dry_run:
                logger.info(f"DRY RUN - Would create issue: {title}")
                logger.debug(f"Issue body:\n{body}")
                self.stats["issues_created"] += 1
                return "dry-run-issue-url"
            
            # Create issue
            issue = self.repo.create_issue(
                title=title,
                body=body,
                labels=self.config.issue_labels,
                assignees=self.config.assignees if self.config.auto_assign else None
            )
            
            # Track this issue
            fingerprint = self.generate_error_fingerprint(error)
            self.reported_fingerprints.add(fingerprint)
            self.fingerprint_timestamps[fingerprint] = time.time()
            self.issue_count_per_hour.append(time.time())
            
            self.stats["issues_created"] += 1
            
            logger.info(f"Created GitHub issue: {issue.html_url}")
            return issue.html_url
            
        except Exception as e:
            self.stats["errors"] += 1
            self.last_issue_error = {
                "timestamp": datetime.now().isoformat(),
                "error_id": error.id,
                "error_type": type(e).__name__,
                "message": str(e),
            }
            if GITHUB_AVAILABLE and isinstance(e, GithubException):
                self.last_issue_error["github_status"] = e.status
                logger.error(
                    "Failed to create GitHub issue (API error %s): %s",
                    e.status,
                    e.data,
                    exc_info=True,
                )
            else:
                logger.error("Failed to create GitHub issue: %s", e, exc_info=True)
            return None
    
    def get_stats(self) -> Dict[str, Any]:
        """
        Get reporter statistics
        
        Returns:
            dict: Statistics
        """
        return {
            **self.stats,
            "config": {
                "enabled": self.config.enabled,
                "repository": self.config.repository,
                "min_error_level": self.config.min_error_level.value,
                "rate_limit_per_hour": self.config.rate_limit_per_hour,
                "dry_run": self.config.dry_run
            },
            "current_rate": len(self.issue_count_per_hour),
            "tracked_fingerprints": len(self.reported_fingerprints),
            "last_issue_error": self.last_issue_error
        }


# Singleton instance
_reporter_instance = None


def get_reporter(config: Optional[IssueReportConfig] = None) -> GitHubIssueReporter:
    """
    Get the singleton GitHub issue reporter instance
    
    Args:
        config: Optional configuration (only used on first call)
        
    Returns:
        GitHubIssueReporter: The reporter instance
    """
    global _reporter_instance
    
    if _reporter_instance is None:
        _reporter_instance = GitHubIssueReporter(config)
    
    return _reporter_instance


def report_error(error: ErrorData) -> Optional[str]:
    """
    Convenience function to report an error
    
    Args:
        error: Error data
        
    Returns:
        Optional[str]: Issue URL if created
    """
    reporter = get_reporter()
    return reporter.create_issue(error)
